src/acquisition/covidcast/migrate_epidata_to_v4.py

```diff
@@ -0,0 +1,188 @@
++------------+
+1 row in set (13 min 49.32 sec)
+
+mysql> select max(id) from epidata.covidcast;
++------------+
+| max(id)    |
++------------+
+| 3740757041 |
++------------+
+1 row in set (0.00 sec)
+
+-- so ~71% coverage of actual rows per allocated ids ( 2647381579 / 3740757041 = .70771278379851347314 )
+
+mysql> select time_value, issue from epidata.covidcast where id=3740757041;
++------------+----------+
+| time_value | issue    |
++------------+----------+
+|   20210927 | 20210930 |
++------------+----------+
+1 row in set (0.01 sec)
+
+mysql> select now();
++---------------------+
+| now()               |
++---------------------+
+| 2022-05-16 16:45:34 |
++---------------------+
+1 row in set (0.00 sec)
+
+'''
+
+
+from delphi.epidata.acquisition.covidcast.database import Database
+import time
+import argparse
+
+def start_tx(cursor):
+  cursor.execute('SET SESSION TRANSACTION ISOLATION LEVEL READ UNCOMMITTED;')
+  cursor.execute('SET autocommit=0;') # starts a transaction as suggested in https://dev.mysql.com/doc/refman/8.0/en/lock-tables.html
+  # NOTE: locks must be specified for any aliases of table names that are used
+  cursor.execute('''LOCK TABLES epidata.covidcast AS cc READ,
+                                signal_load WRITE, signal_load AS sl WRITE,
+                                signal_history WRITE,
+                                signal_latest WRITE,
+                                signal_dim WRITE, signal_dim AS sd READ,
+                                geo_dim WRITE, geo_dim AS gd READ;''')
+  cursor.execute('SET unique_checks=0;')
+
+def finish_tx(cursor):
+  cursor.execute('SET unique_checks=1;')
+  cursor.execute('COMMIT;')
+  cursor.execute('UNLOCK TABLES;')
+
+
+def do_batches(db, start, upper_lim, batch_size):
+  # NOTE: upper_lim is not actually selected for ; make sure it exceeds any ids you want to include
+  batch_lower = start
+
+  while batch_lower < upper_lim:
+    batch_upper = min(batch_lower + batch_size, upper_lim)
+
+    # NOTE: first rows of column names are identical, second rows are for specifying a rename and a literal
+    batch_sql = f"""
+      INSERT INTO signal_load (
+        `issue`, `source`, `signal`, geo_type, geo_value, time_type, time_value, `value`, stderr, sample_size, `lag`, value_updated_timestamp, is_latest_issue, missing_value, missing_stderr, missing_sample_size
+      ) SELECT
+        `issue`, `source`, `signal`, geo_type, geo_value, time_type, time_value, `value`, stderr, sample_size, `lag`, value_updated_timestamp, is_latest_issue, missing_value, missing_stderr, missing_sample_size
+      FROM epidata.covidcast AS cc
+      USE INDEX(`PRIMARY`)
+      WHERE {batch_lower} <= cc.id AND cc.id < {batch_upper}; """
+    # TODO: use LIMIT instead of id range??
+    # TODO: might it be worth adding "ORDER BY id ASC" ?
+
+    if use_transaction_wrappers:
+      start_tx(db._cursor)
+
+    print(f"-=-=-=-=-=-=-=- RUNNING BATCH STARTING AT {batch_lower} -=-=-=-=-=-=-=-")
```

Comment:
Is print okay here since it's the migrate script? Or should this be changed to a log file?

Reply:
I think its ok to keep it as print() because this is only going to be run ~2 times total in the process of migrating the data to v4, and there isnt much to be gained from structured or searchable logs.